earth2grid/healpix.py

```diff
@@ -91,10 +96,57 @@ def pad(x: torch.Tensor, padding: int) -> torch.Tensor:
         return healpixpad.HEALPixPadFunction.apply(x.unsqueeze(2), padding).squeeze(2)
 
 
+def _apply_cuhpx_remap(func, x, **kwargs):
+    shape = x.shape
+    x = x.view(-1, 1, shape[-1])
+    nside = npix2nside(x.shape[-1])
+    x = func(x.contiguous(), **kwargs, nside=nside)
+    x = x.contiguous()
+    x = x.view(shape[:-1] + (-1,))
+    return x
+
+
+def npix2nside(npix: int):
+    nside = math.sqrt(npix // 12)
+    return int(nside)
+
+
+def npix2level(npix: int):
+    return nside2level(npix2nside(npix))
+
+
+def nside2level(nside: int):
+    return int(math.log2(nside))
+
+
 class PixelOrder(Enum):
     RING = 0
     NEST = 1
 
+    def reorder_from_cuda(self, x, src: "PixelOrderT"):
+        if self == PixelOrder.RING:
+            return src.to_ring_cuda(x)
+        elif self == PixelOrder.NEST:
+            return src.to_nest_cuda(x)
+
+    def to_ring_cuda(self, x: torch.Tensor):
+        if self == PixelOrder.RING:
+            return x
+        elif self == PixelOrder.NEST:
+            return _apply_cuhpx_remap(cuhpx.nest2ring, x)
+
+    def to_nest_cuda(self, x: torch.Tensor):
+        if self == PixelOrder.RING:
+            return _apply_cuhpx_remap(cuhpx.ring2nest, x)
+        elif self == PixelOrder.NEST:
+            return x
+
+    def to_xy_cuda(self, x: torch.Tensor, dest: "XY"):
+        if self == PixelOrder.RING:
+            return _apply_cuhpx_remap(cuhpx.ring2flat, x, clockwise=dest.clockwise, origin=dest.origin.name)
+        elif self == PixelOrder.NEST:
```

Comment:
Why no check for `XY`? You could use `xy2xy` which would be nice for arbitrary `XY to `HEALPIX_PAD_XY` and back if needed right?

Reply:
`XY` is a separate class, so has its own versions of these. These methods aren't intended to be public. But are merely implementing a double dispatch pattern to simplify the logic in `earth2grid.healpix.reorder` and `earthgrid.healpix.Grid.reorder`. The cpu/gpu logic is also handled in those two functions.